libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -737,28 +746,62 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
 		_forStatement.loopExpression()->accept(*this);
 
 	m_context.appendJumpTo(loopStart);
+
+	endVisitLoop(&_forStatement);
+
 	m_context << loopEnd;
 
 	m_continueTags.pop_back();
 	m_breakTags.pop_back();
 
+	// For the case where no break is executed:
+	// loop initialization variables have to be freed
+	popBlockScopedVariables(&_forStatement);
+
 	checker.check();
 	return false;
 }
 
 bool ContractCompiler::visit(Continue const& _continueStatement)
 {
-	CompilerContext::LocationSetter locationSetter(m_context, _continueStatement);
-	if (!m_continueTags.empty())
-		m_context.appendJumpTo(m_continueTags.back());
-	return false;
+	return visitBreakContinue(&_continueStatement);
 }
 
 bool ContractCompiler::visit(Break const& _breakStatement)
 {
-	CompilerContext::LocationSetter locationSetter(m_context, _breakStatement);
-	if (!m_breakTags.empty())
+	return visitBreakContinue(&_breakStatement);
+}
+
+unsigned ContractCompiler::loopVariablesSize()
+{
+	solAssert(!m_loops.empty(), "");
+	unsigned sizeOnStack = 0;
+	for (auto _decl : m_loopScopedVariables[m_loops.back()])
+		sizeOnStack += _decl->type()->sizeOnStack();
+	return sizeOnStack;
+}
+
+bool ContractCompiler::visitBreakContinue(Statement const* _statement)
+{
+	CompilerContext::LocationSetter locationSetter(m_context, *_statement);
+	unsigned sizeOnStack = loopVariablesSize();
+	// The context stack needs to be adjusted because the following
+	// pops are extra in the sense that they don't come from a normal
+	// block endVisit.
+	m_context.adjustStackOffset(sizeOnStack);
```

Comment:
It is probably better to move this to the end of the function, because the adjustment has to be made for the dangling block that comes after the unconditional jump.